### ISSUE TITLE
Strange behavior when implicitly coercing Date to POSc

### ISSUE THREAD

**Issue by ProfFancyPants** [Rdatatable/data.table#2392]
I am not sure if this is the expected behavior or not. Either way, it caught me by surprise. 

```r
date_vec <- c('2011-06-02 00:00:00', '39861')
d_ <- d.t(dv = date_vec)
d_[, ld := lubridate::parse_date_time(dv, orders = "y-m-d HMS")] 
d_[!is.na(dv) & is.na(ld), dv_num := as.numeric(dv)]
## Assigns the the as.Date default origin "1970-01-01 03:58:12" with some change.
d_[!is.na(dv_num), ld := as.Date(dv_num, origin = "1899-12-30")]

str(d_$ld)
str(as.Date(d_$dv_num, origin = "1899-12-30"))
## The fix ##
d_[!is.na(dv_num), ld := as.POSIXct(as.Date(dv_num, origin = "1899-12-30"))]
```

Is this the way it is supposed to work? I must admit date-time classes elude me somewhat, but this does seem overly complicated.

**Comment by eantonya** [comment]
I don't really understand what you expected/which part is strange..?

**Comment by MichaelChirico** [comment]
I'm not sure what this has to do with data.table either?

**Comment by MichaelChirico** [comment]
Oh, I see. This indeed differs from base behavior.

In base:

```
library(lubridate)
dv = c('2011-06-02 00:00:00', '39861')
ld = parse_date_time(dv, orders = 'y-m-d HMS')
dv_num = numeric(length(dv))
idx = !is.na(dv) & is.na(ld)
dv_num[idx] = as.numeric(dv[idx])
idx2 = !is.na(dv_num)
ld[idx2] = as.Date(dv_num, origin = '1899-12-30')

str(ld)
# POSIXct[1:2], format: "1899-12-30" "2009-02-17"
```
Somehow `[` here knew to convert the `Date` integer into the corresponding `POSIXct` integer.

In `data.table`, however, the an integer was carried over to the `POSIXct` representation (with some weird time zone math), which is unexpected:

```
str(d_$ld)
# POSIXct[1:2], format: "2011-06-02 00:00:00" "1970-01-01 03:58:12"
```

@ProfFancyPants yes, you should be explicitly converting to `POSIXct` from `Date` before assignment to avoid this.

In R, time objects are integers (stored strangely as `numeric`, see `?IDate`).

By default a `Date` object is the integer number of _days_ since January 1, 1970;

`POSIXct` objects are the integer number of _seconds_ since January 1, 1970 (time-zone dependent).

**Comment by MichaelChirico** [comment]
@ProfFancyPants shot in the dark, this data comes from an Excel file? Are you using `read_excel`, which has a native parser of dates?

**Comment by ProfFancyPants** [comment]
@MichaelChirico that is not the slightest bit a shot in the dark. ;) 

Sorry, I did not explain in more detail; I was a bit rushed for time.  The parsing of files is done within a much larger automated system. At the moment openxlsx::read.xlsx is used for .xlsx and XLConnect::readWorksheet is used for .xls. I didn't go with readxl::read_excel because it was prone to catastrophic failure at the time. 

I understand your explanation, Michael. Thank you for taking the time to explain. 

> In data.table, however, the an integer was carried over to the POSIXct representation (with some weird time zone math), which is unexpected

There are still some thoughts I want to address. The sticking point is how data.table implicitly converts values when being set within a vector of a particular class. This is like setting a numeric value inside of an integer column:
```r
d_ <- data.table(iv = 1L:10L)
d_[between(iv, 4.5, 6.5), iv := 0.5]
# Warning in `[.data.table`(d_, between(iv, 4.5, 6.5), `:=`(iv, 0.5)) :
# Coerced 'double' RHS to 'integer' to match the column's type; may have truncated precision. 
# Either change the target column ['iv'] to 'double' first (by creating a new 'double' vector length 
# 10 (nrows of entire table) and assign that; i.e. 'replace' column), or coerce RHS to 'integer' (e.g. 1L, 
# NA_[real|integer]_, as.*, etc) to make your intent clear and for speed. Or, set the column type 
# correctly up front when you create the table and stick to it, please.
# Calls: [ -> [.data.table
```

The reason the warning is signalled is because precision is lost due to the coerision to an integer. When there is no precision loss, the set implicit:
```r
d_ <- data.table(iv = 1L:10L)
d_[between(iv, 4.5, 6.5), iv := as.numeric(0)]
```

As you pointed out, Micheal, the same thing is happening with regard to `Date` class being coerced into `POSIXct`, however, in this case, it is two doubles being implicitly interpreted as perfectly transferable. My issue is, who on their life would this on purpose?!  I understand that deep down, they are the same double class:
```r
as.Date("1970-01-02 00:00:01") %>% dputc
# structure(1,class="Date")
as.POSIXct("1970-01-02 00:00:01") %>% dputc
# structure(104401,class=c("POSIXct","POSIXt"),tzone="")
```
But shouldn't there be some sort of warning when this is happening? Just for transfers between date /time classes at least?

**Comment by MichaelChirico** [comment]
Yes, I absolutely agree it's a bug, and should probably error in this case (or perhaps warn or perhaps even offer automated conversion)
